Replace C*Pos intermediate variables with expressions

gizatt opened the pull request:
This replaces some continuous variables that were carried around to make writing the rotation constraints easier. Instead, they are now formed as expressions, simplifying the optimization.

This has been deployed on my stuff for a while, but the timing differences it makes on tests relying on the rotations show that it does increase setup time, but makes longer-running solves more efficient:

All times reported below are found by running `time <the test binary>` and taking the user time (which I interpret as total CPU time spent for the process running non-syscall code). I tested each case 3 times, and numbers didn't change more than a few percent. The first two cases have lots of short-running solves, and last two cases have much longer-running solves.

| Test | Total CPU core time without change   | Total CPU core time with change |
| -------------- | ------------- | ------------- |
| rotation_constraint_test (all) | 1m 7s  | 1m 24s  |
| TestMcCormick | 0m 1.4s  | 0m 2.0s  |
| rotation_constraint_limit_test (all) | 37m 0s | 29-33m 0s |
| global_inverse_kinematics_collision_avoidance_test | 10m 40s | 4m 10s |

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6533)
<!-- Reviewable:end -->

hongkai-dai commented:
+@hongkai-dai  for feature review

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-:-KoTa4on5-AGpOoezZcz:b-8zd4a7)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/solvers/rotation_constraint.cc, line 921 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoTbqlxw4juD8xvdntO:-KoTbqlxw4juD8xvdntP:b-3tcgqa) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L921)):*
> ```cpp
>   // pass into the constraints changes for the k=N-1 case.  Otherwise we
>   // could do a simple substitution everywhere.
>   // TODO(russt): Use symbolic constraints and remove these decision variables!
> ```

Remove this TODO?

---

*[drake/solvers/rotation_constraint.cc, line 929 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoTcA0UqV0kW4x5fesA:-KoTcA0UqV0kW4x5fesB:b-f9gtl) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L929)):*
> ```cpp
>     BRneg.push_back(
>         prog->NewBinaryVariables<3, 3>("BRneg" + std::to_string(k)));
>     CRpos.push_back(Eigen::Matrix3d::Ones());
> ```

Why we need to push back ones here? I think we can just do 
```C++
CRpos.reserve(num_binary_vars_per_half_axis);
CRneg.reserve(num_binary_vars_per_half_axis);
for (int k = 0; k < num_binary_vars_per_half_axis; k++) {
    BRpos.push_back(
        prog->NewBinaryVariables<3, 3>("BRpos" + std::to_string(k)));
    BRneg.push_back(
        prog->NewBinaryVariables<3, 3>("BRneg" + std::to_string(k)));
}
for (int k = 0; k < num_binary_vars_per_half_axis - 1; k++) {
  CRpos.push_back(BRpos[k] - BRpos[k + 1]);
  CRneg.push_back(BRneg[k] - BRneg[k + 1]);
}
CRpos.push_back(BRpos[num_binary_vars_per_half_axis - 1].cast<symbolic::Expression>());
CRneg.push_back(BRneg[num_binary_vars_per_half_axis - 1].cast<symbolic::Expression>());
```
And remove the changes on `CRpos, CRneg` down below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot mac-clang-bazel-experimental please

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-:-KoUHmT9l1Kl2yZDXkyO:bjwebpu)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/solvers/rotation_constraint.cc, line 569 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUY3Q1DYncgNn0NTFB:-KoUY3Q1DYncgNn0NTFC:b-o7mjrv) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L569)):*
> ```cpp
>     MathematicalProgram* prog, const VectorDecisionVariable<3>& v,
>     const std::vector<Eigen::Matrix<Expression, 3, 1>>& cpos,
>     const std::vector<Eigen::Matrix<Expression, 3, 1>>& cneg,
> ```

BTW, we have `Vector3<Expression>`, if we include `drake/common/eigen_types.h`.

---

*[drake/solvers/rotation_constraint.cc, line 585 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUXwgoL7QZY5XLfgVY:-KoUXwgoL7QZY5XLfgVZ:b-7ropln) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L585)):*
> ```cpp
>         box_max(2) = EnvelopeMinValue(zi + 1, N);
> 
>         Eigen::Matrix<Expression, 3, 1> this_cpos, this_cneg;
> ```

ditto

---

*[drake/solvers/rotation_constraint.cc, line 635 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYIbXRsPb4IzrtLFj:-KoUYIbXRsPb4IzrtLFk:b-7ropln) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L635)):*
> ```cpp
>             }
>             Eigen::Vector3d orthant_u;
>             Eigen::Matrix<Expression, 3, 1> orthant_c;
> ```

ditto

---

*[drake/solvers/rotation_constraint.cc, line 691 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYHabEsshn2Y3NYn-:-KoUYHabEsshn2Y3NYn0:b-7ropln) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L691)):*
> ```cpp
> 
>             Eigen::Vector3d orthant_normal;
>             Eigen::Matrix<Expression, 3, 1> orthant_c;
> ```

ditto

---

*[drake/solvers/rotation_constraint.cc, line 759 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYgnXRA0g-s4UutNd:-KoUYgnXRA0g-s4UutNe:btm8kaw) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L759)):*
> ```cpp
>               //   we can impose a tighter Lorentz cone constraint
>               //     [|sin(theta)|, n'*v1, n'*v2] is in the Lorentz cone.
>               a << orthant_normal, -2, -2, -2;
> ```

`a` is not used anymore?

If setup time becomes a concern, we will avoid this symbolic constraint. We do need to parse the same symbolic expression for twice here, one for >=, and one for <= .

---

*[drake/solvers/rotation_constraint.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUZ9ObMasOlZo3TCCo:-KoUZ9ObMasOlZo3TCCp:b-6390mj) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.h#L95)):*
> ```objc
> 
> using AddRotationMatrixMcCormickEnvelopeReturnType =
> std::tuple<std::vector<Eigen::Matrix<drake::symbolic::Expression, 3, 3>>,
> ```

BTW, we also have `Matrix3<symbolic::Expression>`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533)*
<!-- Sent from Reviewable.io -->

gizatt commented:
Addressed the comments as much as possible, waiting on your PR to go through to look deeper at the failing test.

---

Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/solvers/rotation_constraint.cc, line 569 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUY3Q1DYncgNn0NTFB:-KoiIh9Do054qkbjpvxS:b-evgdm) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L569)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

BTW, we have `Vector3<Expression>`, if we include `drake/common/eigen_types.h`.
</blockquote></details>

Thanks, switched over to that!

---

*[drake/solvers/rotation_constraint.cc, line 585 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUXwgoL7QZY5XLfgVY:-KoiIqpkWh34YPpRy13_:b-z6fzt4) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L585)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Switched

---

*[drake/solvers/rotation_constraint.cc, line 635 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYIbXRsPb4IzrtLFj:-KoiIzg6o-mnmO0ZNSZR:b-z6fzt4) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L635)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Switched

---

*[drake/solvers/rotation_constraint.cc, line 691 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYHabEsshn2Y3NYn-:-KoiJ-fozNcPTqHnQF1l:b-z6fzt4) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L691)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Switched

---

*[drake/solvers/rotation_constraint.cc, line 759 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYgnXRA0g-s4UutNd:-KoiK_Z1e1h6RRp9C6RI:b584hao) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L759)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`a` is not used anymore?

If setup time becomes a concern, we will avoid this symbolic constraint. We do need to parse the same symbolic expression for twice here, one for >=, and one for <= .
</blockquote></details>

Zapped, would it help if I stored the shared term in its own object (rather than repeating)? Or would the parsing not take advantage of that?

---

*[drake/solvers/rotation_constraint.cc, line 921 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoTbqlxw4juD8xvdntO:-KoiLx65Z-wI-_NN92x6:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L921)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Remove this TODO?
</blockquote></details>

Done

---

*[drake/solvers/rotation_constraint.cc, line 929 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoTcA0UqV0kW4x5fesA:-KoiNSpxRu-VLEv_XWYx:b901njl) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L929)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Why we need to push back ones here? I think we can just do 
```C++
CRpos.reserve(num_binary_vars_per_half_axis);
CRneg.reserve(num_binary_vars_per_half_axis);
for (int k = 0; k < num_binary_vars_per_half_axis; k++) {
    BRpos.push_back(
        prog->NewBinaryVariables<3, 3>("BRpos" + std::to_string(k)));
    BRneg.push_back(
        prog->NewBinaryVariables<3, 3>("BRneg" + std::to_string(k)));
}
for (int k = 0; k < num_binary_vars_per_half_axis - 1; k++) {
  CRpos.push_back(BRpos[k] - BRpos[k + 1]);
  CRneg.push_back(BRneg[k] - BRneg[k + 1]);
}
CRpos.push_back(BRpos[num_binary_vars_per_half_axis - 1].cast<symbolic::Expression>());
CRneg.push_back(BRneg[num_binary_vars_per_half_axis - 1].cast<symbolic::Expression>());
```
And remove the changes on `CRpos, CRneg` down below.
</blockquote></details>

Good point, made this modification and moved some blocks around to try to clean up the organization a bit to more clearly separate BR* and CR* setup.

---

*[drake/solvers/rotation_constraint.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUZ9ObMasOlZo3TCCo:-KoiJS0pZBJGEE6XDKJa:b-z6fzt4) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.h#L95)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

BTW, we also have `Matrix3<symbolic::Expression>`.
</blockquote></details>

Switched

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-:-Koibl3XnTlTVxH79xtZ:b-9mu2d7)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think after my PR is merged, you will still have a failing test (the error is smaller, but still above the threshold). I would suggest to increase that threshold to 0.22 in https://github.com/RobotLocomotion/drake/blob/master/drake/multibody/dev/test/global_inverse_kinematics_reachable_test.cc#L35, after my PR is merged.

+@sammy-tri for platform review please.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/rotation_constraint.cc, line 759 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-KoUYgnXRA0g-s4UutNd:-KoifUp0pOyToeeoaVnh:bqejyuo) ([raw file](https://github.com/robotlocomotion/drake/blob/1a9c71dc67eadc8abbe118df50384304663ca922/drake/solvers/rotation_constraint.cc#L759)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Zapped, would it help if I stored the shared term in its own object (rather than repeating)? Or would the parsing not take advantage of that?
</blockquote></details>

You will be able to store the shared term `orthant_normal.dot(v1) - 2 * orthant_c.sum()` into a `Polynomial` object, when #6545 is merged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-:-Koig69XFAFl4JY82ogG:b-4pmx8c)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6533#-:-KomLcG8ZbeBeyPWH_L3:bnfp4nl)*
<!-- Sent from Reviewable.io -->
